File: google-news-scraper/browser/nodejs/google-news.mjs
// GoogleNews scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Target surface: `https://news.google.com/search?q=<query>&hl=en` (and any
// topic page of the same shape). The scraper renders the SPA, waits for the
// article cards to mount, and extracts one row per `<article>` tag.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 180;
// Cards are rendered as `<div class="m5k28">` wrappers (sometimes inside
// `<article>` tags on older variants). We wait for the headline anchor
// class `JtKRv` which is stable across both layouts.
const READY_SELECTOR = "a.JtKRv, article a";

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function fetchRenderedHtml(url, readySelector, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 2 } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      await page.setViewport({ width: 1366, height: 900 });
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
      try { await page.waitForSelector(readySelector, { timeout: 20000 }); } catch (_) {}
      // Give the SPA a beat to fan out card thumbnails.
      await new Promise((r) => setTimeout(r, 2000));
      const html = await page.content();
      if (html && html.includes("JtKRv")) return html;
      lastError = new Error("no article headlines in HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
    if (attempt < retries) await new Promise((r) => setTimeout(r, 3000 + attempt * 2000));
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

// ---------------- scrape functions ----------------

export async function scrapeNews(query) {
  const url = `https://news.google.com/search?q=${encodeURIComponent(query)}&hl=en-US&gl=US&ceid=US:en`;
  const html = await fetchRenderedHtml(url, READY_SELECTOR);
  return parseNews(html);
}

export async function scrapeTopic(topicId) {
  const url = `https://news.google.com/topics/${topicId}?hl=en-US&gl=US&ceid=US:en`;
  const html = await fetchRenderedHtml(url, READY_SELECTOR);
  return parseNews(html);
}

// ---------------- parser ----------------

export function parseNews(html) {
  const $ = cheerio.load(html);
  const seen = new Set();
  const out = [];
  let position = 0;
  // Iterate over the visible headline anchors. Each `a.JtKRv` is rendered
  // once per card and carries an `aria-label` of the form:
  //   "<title> - <source> - <time> - By <author>"
  // which is by far the cleanest way to recover the per-card fields.
  $("a.JtKRv").each((_, el) => {
    const a = $(el);
    const href = a.attr("href") || "";
    const titleText = normalise(a.text());
    if (!href || !titleText) return;
    const absUrl = href.startsWith("./") ? `https://news.google.com${href.slice(1)}` : href;
    if (seen.has(absUrl)) return;
    seen.add(absUrl);

    const aria = a.attr("aria-label") || "";
    const [, source = "", time = ""] = splitAriaLabel(aria, titleText);

    // Walk up to the card container to pick up the thumbnail.
    let thumbnail = "";
    const card = a.closest("div.m5k28, article, div.IBr9hb, div.XlKvRb");
    if (card.length) {
      card.find("img").each((__, img) => {
        const src = $(img).attr("src") || $(img).attr("data-src") || "";
        if (/^https?:/.test(src)) { thumbnail = src; return false; }
        return undefined;
      });
    }

    position += 1;
    out.push({
      position,
      title: titleText,
      url: absUrl,
      source: source,
      time: time,
      thumbnail: thumbnail,
    });
  });
  return out;
}

// aria-label format: "<title> - <source> - <time> - By <author>"
// where `time` is a relative phrase ("3 hours ago", "Yesterday", "2 days ago", …).
// Returns ["", source, time].
function splitAriaLabel(aria, title) {
  if (!aria) return ["", "", ""];
  // Trim the leading title (may differ by a stray punctuation; do a permissive prefix match).
  let rest = aria;
  if (aria.startsWith(title)) {
    rest = aria.slice(title.length);
  } else {
    const idx = aria.indexOf(" - ");
    if (idx >= 0) rest = aria.slice(idx);
  }
  rest = rest.replace(/^ -\s*/, "");
  const parts = rest.split(" - ");
  return ["", normalise(parts[0] || ""), normalise(parts[1] || "")];
}

function normalise(s) {
  return String(s || "").replace(/\s+/g, " ").trim();
}

File: google-news-scraper/browser/python/google_news.py
"""GoogleNews scraper using the official Scrapeless Python SDK + Playwright over CDP.

Target surface: ``https://news.google.com/search?q=<query>&hl=en`` and any
topic page of the same shape. The scraper renders the SPA, waits for the
headline anchors (``a.JtKRv``) to mount, and yields one ``Article`` per
visible card.
"""

from __future__ import annotations

import asyncio
import os
import re
from dataclasses import asdict, dataclass
from typing import Any
from urllib.parse import quote_plus

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 180
# Cards render as `<div class="m5k28">` (sometimes inside `<article>` on
# older layouts). We wait for the headline anchor `a.JtKRv` which is stable
# across both layouts.
READY_SELECTOR = "a.JtKRv, article a"


@dataclass
class Article:
    position: int
    title: str
    url: str
    source: str = ""
    time: str = ""
    thumbnail: str = ""


def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()


async def _fetch_rendered_html(
    url: str, ready_selector: str = READY_SELECTOR, *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY, retries: int = 2,
) -> str:
    last_error: Exception | None = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
            try:
                page = await browser.new_page(viewport={"width": 1366, "height": 900})
                await page.goto(url, wait_until="domcontentloaded", timeout=45000)
                try:
                    await page.wait_for_selector(ready_selector, timeout=20000)
                except Exception as e:
                    logger.warning("ready selector not seen (continuing): {}", e)
                await asyncio.sleep(2)
                html = await page.content()
                if html and "JtKRv" in html:
                    return html
                last_error = RuntimeError("no headline anchors in HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
        if attempt < retries:
            await asyncio.sleep(3 + attempt * 2)
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")


# ---------------------------------------------------------------------------
# Scrape functions
# ---------------------------------------------------------------------------


async def scrape_news(query: str) -> list[Article]:
    url = f"https://news.google.com/search?q={quote_plus(query)}&hl=en-US&gl=US&ceid=US:en"
    html = await _fetch_rendered_html(url)
    return parse_news(html)


async def scrape_topic(topic_id: str) -> list[Article]:
    url = f"https://news.google.com/topics/{topic_id}?hl=en-US&gl=US&ceid=US:en"
    html = await _fetch_rendered_html(url)
    return parse_news(html)


# ---------------------------------------------------------------------------
# Parser
# ---------------------------------------------------------------------------


_WS_RE = re.compile(r"\s+")
_HTTP_RE = re.compile(r"^https?:")


def _normalise(text: str) -> str:
    return _WS_RE.sub(" ", text or "").strip()


def _split_aria_label(aria: str, title: str) -> tuple[str, str]:
    """aria-label format: "<title> - <source> - <time> - By <author>".

    Returns (source, time).
    """
    if not aria:
        return "", ""
    if aria.startswith(title):
        rest = aria[len(title):]
    else:
        idx = aria.find(" - ")
        rest = aria[idx:] if idx >= 0 else aria
    rest = re.sub(r"^ -\s*", "", rest)
    parts = rest.split(" - ")
    return _normalise(parts[0] if parts else ""), _normalise(parts[1] if len(parts) > 1 else "")


def parse_news(html: str) -> list[Article]:
    sel = Selector(text=html)
    seen: set[str] = set()
    out: list[Article] = []
    position = 0
    for a in sel.css("a.JtKRv"):
        href = a.attrib.get("href", "")
        title = _normalise("".join(a.css("*::text").getall()))
        if not href or not title:
            continue
        abs_url = f"https://news.google.com{href[1:]}" if href.startswith("./") else href
        if abs_url in seen:
            continue
        seen.add(abs_url)

        aria = a.attrib.get("aria-label", "")
        source, time = _split_aria_label(aria, title)

        # Walk up to the card container for the thumbnail.
        thumbnail = ""
        cards = a.xpath(
            "ancestor::*[contains(concat(' ', normalize-space(@class), ' '), ' m5k28 ')"
            " or self::article"
            " or contains(concat(' ', normalize-space(@class), ' '), ' IBr9hb ')"
            " or contains(concat(' ', normalize-space(@class), ' '), ' XlKvRb ')][1]"
        )
        if cards:
            for img in cards[0].css("img"):
                src = img.attrib.get("src") or img.attrib.get("data-src") or ""
                if _HTTP_RE.match(src):
                    thumbnail = src
                    break

        position += 1
        out.append(Article(
            position=position,
            title=title,
            url=abs_url,
            source=source,
            time=time,
            thumbnail=thumbnail,
        ))
    return out


def to_dict(obj: Any) -> Any:
    if hasattr(obj, "__dataclass_fields__"):
        return asdict(obj)
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj
